crypto/zk/circuit/circuit.go

```diff
@@ -0,0 +1,278 @@
+package circuit
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+
+	"go.vocdoni.io/dvote/log"
+)
+
+// By default set the circuit base dir empty which means that the artifacts will
+// be downloaded or loaded from the LocalDir path associated to the circuit config
+var circuitsBaseDir = ""
+var BaseDir = &circuitsBaseDir
+
+// SetBaseDir allows to modify the default base dir to download and load the
+// circuits artifacts. They will be dowloaded or loaded from the LocalDir path
+// associated to the circuit config but inside of BaseDir folder path.
+func SetBaseDir(dir string) {
+	BaseDir = &dir
+}
+
+var downloadCircuitsTimeout = time.Minute * 5
+
+// ZkCircuit struct wraps the circuit configuration and contains the file
+// content of the circuit artifacts (provingKey, verificationKey and wasm)
+type ZkCircuit struct {
+	ProvingKey      []byte
+	VerificationKey []byte
+	Wasm            []byte
+
+	Config ZkCircuitConfig
+}
+
+// LoadZkCircuitByTag gets the circuit configuration associated to the provided
+// tag or gets the default one and load its artifacts to prepare the circuit to
+// be used.
+func LoadZkCircuitByTag(configTag string) (*ZkCircuit, error) {
+	circuitConf := GetCircuitConfiguration(configTag)
+
+	ctx, cancel := context.WithTimeout(context.Background(), downloadCircuitsTimeout)
+	defer cancel()
+
+	zkCircuit, err := LoadZkCircuit(ctx, circuitConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return zkCircuit, nil
+}
+
+// LoadZkCircuit function load the circuit artifacts based on the configuration
+// provided. First, tries to load the artifacts from local storage, if they are
+// not available, tries to download from their remote location. Then,
+func LoadZkCircuit(ctx context.Context, config ZkCircuitConfig) (*ZkCircuit, error) {
+	// Join the local base path with the local dir set up into the circuit
+	// configuration
+	config.LocalDir = filepath.Join(*BaseDir, config.LocalDir)
+	circuit := &ZkCircuit{Config: config}
+
+	// load the artifacts of the provided circuit from the local storage
+	if err := circuit.LoadLocal(); err == nil {
+		// tries to verify the loaded artifacts, if it success, returns the
+		// loaded circuit, else continue.
+		correct, err := circuit.VerifiedCircuitArtifacts()
+		if err == nil && correct {
+			return circuit, nil
+		}
+	}
+
+	// if the circuit is not available locally, tries to download from its
+	// remote location
+	if err := circuit.LoadRemote(ctx); err != nil {
+		return nil, err
+	}
+
+	// checks hashes of current files
+	if correct, err := circuit.VerifiedCircuitArtifacts(); err != nil {
+		return nil, err
+	} else if !correct {
+		return nil, fmt.Errorf("download artifacts does not match with the expected ones")
```

Comment:
"hashes from downloaded artifacts don't match the expected ones"

Reply:
updated